[feature] new GBT phase finding algo

jsturdy opened the pull request:
## Description
Addressing #248, supersedes #249 

Now also supersedes https://github.com/cms-gem-daq-project/vfatqc-python-scripts/pull/254

### Types of changes
- [x] New feature (non-breaking change which adds functionality)

## Motivation and Context
Incorporate information about the GBT phases as well as real-world data on different scans.
Collapses the possible cases into 2:
* `0` bad phases found: use method developed by @bdorney 
* `> 0` bad phases found: use a window search around each bad phase to find the window with the most good phase counts, and set the phase in that window (window and set point currently defaulted to 4, but can be changed independently)

## How Has This Been Tested?
Tested on sample data from QC8 test that showed new bad phase cases, probably needs live testing for final sign-off
```
Ideal phase for VFAT0 is:  [7] [50. 35. 47.  0. 50. 50. 50. 50. 50. 50. 50.  0. 50. 50. 50.  0.]
Ideal phase for VFAT1 is:  [9] [50. 35. 48. 50. 26.  2. 50. 50. 50. 50. 50. 50.  0. 50. 50.  0.]
Ideal phase for VFAT2 is:  [5] [50. 35. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50.  0.]
Ideal phase for VFAT3 is:  [6] [17. 35. 49. 50. 50. 50. 50. 50. 50.  0. 50. 50. 50. 50. 50.  0.]
Ideal phase for VFAT4 is:  [6] [47.  0. 49. 50. 50. 50. 50. 50. 50.  0. 50. 50. 50. 50. 50.  0.]
Ideal phase for VFAT5 is:  [12] [13. 35. 49. 50. 50. 50. 50. 50.  0. 50. 50. 50. 50. 50. 50.  0.]
Ideal phase for VFAT6 is:  [8] [46. 35. 49. 50.  0. 50. 50. 50. 50. 50. 50. 50.  0. 50. 50.  0.]
Ideal phase for VFAT7 is:  [7] [48. 35. 48.  0. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50.  0.]
Ideal phase for VFAT8 is:  [7] [46. 35. 46.  0. 50. 50. 50. 50. 50. 50. 50.  0. 50. 50. 50.  0.]
Ideal phase for VFAT9 is:  [6] [50. 35. 47. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50.  0.]
Ideal phase for VFAT10 is:  [11] [49. 35. 47. 50. 50. 50. 50. 40. 50. 50. 50. 50. 50. 50. 50.  0.]
Ideal phase for VFAT11 is:  [12] [17. 35. 47. 50. 50. 50. 50. 50.  0. 50. 50. 50. 50. 50. 50.  0.]
Ideal phase for VFAT12 is:  [5] [45.  0. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50.  0.]
Ideal phase for VFAT13 is:  [6] [46. 35. 49. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50.  0.]
Ideal phase for VFAT14 is:  [10] [36. 35. 49. 50. 50. 50.  0. 50. 50. 50. 50. 50. 50. 50.  0.  0.]
Ideal phase for VFAT15 is:  [6] [42. 35.  5. 50. 50. 50. 50. 50. 50. 50.  0. 50. 50. 50. 50.  0.]
Ideal phase for VFAT16 is:  [5] [43. 35. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50.  0.]
Ideal phase for VFAT17 is:  [9] [46. 35. 50. 50. 50.  0. 50. 50. 50. 50. 50. 50. 50.  0. 50.  0.]
Ideal phase for VFAT18 is:  [8] [48. 35. 50. 50.  0. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50.  0.]
Ideal phase for VFAT19 is:  [5] [48. 35. 50. 50. 50. 50. 50. 50. 50. 50.  0. 50. 50. 50. 50.  0.]
Ideal phase for VFAT20 is:  [10] [49. 35. 50. 50. 50. 50.  0. 50. 50. 50. 50. 50. 50.  0. 50.  0.]
Ideal phase for VFAT21 is:  [9] [49. 35. 50. 50. 50.  0. 50. 50. 50. 50. 50. 50. 50.  0. 50.  0.]
Ideal phase for VFAT22 is:  [5] [49. 35. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50.  0.]
Ideal phase for VFAT23 is:  [7] [49. 35. 48.  0. 50. 50. 50. 50. 50. 50. 50.  0. 50. 50. 50.  0.]
```

## Before merging
Once tested, the old algo block can possibly be removed (and maybe any associated additional files that were old-algo specific).
If approved, this can be done during a `rebase`

jsturdy commented on testConnectivity.py:
```diff
@@ -87,10 +92,12 @@ def scaCommIsGood(amc, maxIter=5, ohMask=0xfff, nOHs=12):
     from reg_utils.reg_interface.common.jtag import initJtagRegAddrs
     initJtagRegAddrs()
     for trial in range(0,maxIter):
+        writeRegister(amc,"GEM_AMC.SLOW_CONTROL.SCA.ADC_MONITORING.MONITORING_OFF",0x0)
```
Does this need to happen inside the loop?

bdorney replied:
It seems your fork point is not up-to-date with my remote branch; this should be writing `0xfff` as illustrated [here](https://github.com/cms-gem-daq-project/vfatqc-python-scripts/pull/249/files#diff-49db7e1b7f530db9c88c1268571e14eeR100)

bdorney replied:
> Does this need to happen inside the loop?

No it can be done before line 99 but after line 98.

bdorney commented:
Your branch point doesn't match my remote.

jsturdy commented:
I branched before you rebased, so since I had made my changes before you rebased, I rebased my changes on the point you rebased from, not on your original (rebased) branch, but since you haven't committed anything since I branched (right?), this shouldn't matter (unless part of your `rebase` involved a change, which it looks like it did)

I've moved the line before the loop, and changed it to the correct value.
To confirm, this is a "turn it off, leave it off" desired behaviour, which is what is currently being done, rather than a "turn if off, revert when done" behaviour

bdorney commented:
> To confirm, this is a "turn it off, leave it off" desired behaviour, which is what is currently being done, rather than a "turn if off, revert when done" behaviour

Yeah this is a "turn it off and leave it off" so that we know the state of the system.  When this block gets ripped out of the `GEM_AMC` FW we'll need to remove these lines as well.

Sorry for the git foo confusion.

bdorney commented on testConnectivity.py:
```diff
@@ -409,75 +436,157 @@ def testConnectivity(args):
+                allBadPhases = np.where(phaseCounts!=args.nPhaseScans)[0]
+                badPhaseCounts = np.delete(allBadPhases,np.where(allBadPhases==15)[0]) ## remove 15 from the list of bad phases
+                phaseSum = 0
+                if len(badPhaseCounts) == 0:
```
~~maybe I misunderstood but I thought this block was being removed in favor of the window method around a dead phase?~~

bdorney replied:
~~I don't see the block that simplifies this...maybe I missed it though; or is there some more git foo confusion going on?~~

bdorney replied:
Nevermind I see the `True pass statement`

bdorney commented:
This seems to still be not even with #249:

```
Step 4: Checking VFAT Communication
Checking GBT Communication (After Programming FPGA)
GBT Communication Is Stil Good

Scanning GBT Phases, this may take a moment please be patient
Phase Scan Results for OH4
  Phase    VFAT0    VFAT1    VFAT2    VFAT3    VFAT4    VFAT5    VFAT6    VFAT7    VFAT8    VFAT9    VFAT10    VFAT11    VFAT12    VFAT13    VFAT14    VFAT15    VFAT16    VFAT17    VFAT18    VFAT19    VFAT20    VFAT21    VFAT22    VFAT23

  0       50       50       50        0       50        0        0       50       50       50         0        50        50        50        50        50        50        50        50        50        50        50         0        50
  1       50       50       50       50       47       50       50       50       50       50        50        47        50        50        50        50        50        50        50        50        50        50        50        50
  2       50       50       50       50       50       50       50       50       50       50        50        50         0        50        50        50        50        50        50         0        50        50        50        50
  3        0       50       50       50       50       50       24       50        0       50        50        50        50        50        50        50        50        50        50        50        50        50        50         0
  4       50       50       50       50       50       50        0       50       50       50        50        50        50        50        50        50        50        50         0        50        50        50        50        50
  5       50       17       50       50       50       50        0       50       50        0        50        50        50        50        50        50        50         1        50        50        50         0        50        50
  6       50       50       35       50       50       50        0       50       50       50        50        50        50        50         0        50        50        50        50        50        50        50        50        50
  7       50       50       50       50       50       50        0       50       50       50        50        50        50        50        50        50        50        50        50        50        50        50        50        50
  8       50       50       50       50       50        0       50       50       50       50         0        50        50        50        50        50        50        50        50        50        50        50         0        50
  9       50       50       50        0        0       50       50       50       50       50        50         0        50        50        50        50        50        50        50        50        50        50        50        50
 10       50       50       50       50       50       50       50       50        0       50        50        50         0        50        50         0        50        50        50         0        50        50        50        50
 11        0       50       50       50       50       50       29       50        0       50        50        50        50        50        50        50        50        50        50        50        50        50        50         0
 12       50       50       50       50       50       50       22        0       50       50        50        50        50        50        50        50        50        50        50        50        50        50        50        50
 13       50        0       50       50       50       50        0       50       50        0        50        50        50        50        50        50        43         0        50        50        50         0        50        50
 14       50       50        0       50       50       50        0       50       50       50        50        50        50        50         0        50        50        50        50        50        50        50        50        50
 15        0        0        0        0        0        0        0        0        0        0         0         0         0         0         0         0         0         0         0         0         0         0         0         0
Traceback (most recent call last):
  File "/opt/cmsgemos/bin/testConnectivity.py", line 1191, in <module>
    testConnectivity(args)
  File "/opt/cmsgemos/bin/testConnectivity.py", line 510, in testConnectivity
    frange  = crange(int(badPhase+1),
NameError: global name 'crange' is not defined
```

The QC7 team in the lab have the above error so it seems like an import is not provided.

bdorney commented on testConnectivity.py:
```diff
@@ -409,75 +436,157 @@ def testConnectivity(args):
+                    for b in badPhaseCounts:
+                        WINDOW = 4 ## Good phase search window
+                        SHIFT  = 4 ## Phase shift from bad phase to set
+                        frange  = crange(int(badPhase+1),
```
missing import on `crange`

bdorney replied:
Also `badPhase` causes a `nameError` I guess this is supposed to be the array index `b`?

bdorney commented on testConnectivity.py:
```diff
@@ -409,75 +436,157 @@ def testConnectivity(args):
+                        bsum = sum(phaseCounts.take(brange, mode='wrap')) # backward sum
+                        tmpPhase = 15
+                        if fsum > psum:
+                            psum     = tmpfsum
```
I guess this is supposed to be `fsum`, otherwise `NameError`

bdorney commented on testConnectivity.py:
```diff
@@ -409,75 +436,157 @@ def testConnectivity(args):
+                            psum     = tmpfsum
+                            tmpPhase = int((badPhase+SHIFT)%GBT_PHASE_RANGE)
+                        if bsum > psum:
+                            psum     = tmpbsum
```
same?

bdorney commented:
Fixing the bugs I pointed out above we have:

```
/data/bigdisk/GEM-Data-Taking/GE11_QC8//GE11-X-S-PAK-0007/connectivityLog_GE11-X-S-PAK-0007_2019.04.30.19.51.log
```

Here the algorithm tries to write a phase for VFAT6 but it choses a bad phase:

```
====================
Phase Scan Results for OH4
====================
  Phase    VFAT0    VFAT1    VFAT2    VFAT3    VFAT4    VFAT5    VFAT6    VFAT7    VFAT8    VFAT9    VFAT10    VFAT11    VFAT12    VFAT13    VFAT14    VFAT15    VFAT16    VFAT17    VFAT18    VFAT19    VFAT20    VFAT21    VFAT22    VF
AT23
-------  -------  -------  -------  -------  -------  -------  -------  -------  -------  -------  --------  --------  --------  --------  --------  --------  --------  --------  --------  --------  --------  --------  --------  ----
----
      0       50       50       50        0       50        0        0       50       50       50         0        50        50        50        50        50        50        50        50        50        50        50         0      
  50
      1       50       50       50       50       50       50       50       50       50       50        50        45        50        50        50        50        50        50        50        50        50        50        50      
  50
      2       50       50       50       50       50       50       50       50       50       50        50        50         0        50        50        50        50        50        50         0        50        50        50      
  50
      3        0       50       50       50       50       50       21       50        0       50        50        50        50        50        50        50        50        50        50        50        50        50        50      
   0
      4       50       50       50       50       50       50        0       50       50       50        50        50        50        50        50        50        50        50         0        50        50        50        50      
  50
      5       50        9       50       50       50       50        0       50       50        0        50        50        50        50        50        50        50         0        50        50        50         0        50      
  50
      6       50       50       25       50       50       50        0       50       50       50        50        50        50        50         0        50        50        50        50        50        50        50        50      
  50
      7       50       50       50       50       50       50        0       50       50       50        50        50        50        50        50        50        50        50        50        50        50        50        50      
  50
      8       50       50       50       50       50        0       50       50       50       50         0        50        50        50        50        50        50        50        50        50        50        50         0      
  50
      9       50       50       50        0        0       50       50       50       50       50        50         0        50        50        50        50        50        50        50        50        50        50        50      
  50
     10       50       50       50       50       50       50       50       50        0       50        50        50         0        50        50         0        50        50        50         0        50        50        50      
  50
     11        0       50       50       50       50       50       25       50        0       50        50        50        50        50        50        50        50        50        50        50        50        50        50      
   0
     12       50       50       50       50       50       50       22        0       50       50        50        50        50        50        50        50        50        50        50        50        50        50        50      
  50
     13       50        0       50       50       50       50        0       50       50        0        50        50        50        50        50        50        48         0        50        50        50         0        50      
  50
     14       50       50        0       50       50       50        0       50       50       50        50        50        50        50         0        50        50        50        50        50        50        50        50      
  50
     15        0        0        0        0        0        0        0        0        0        0         0         0         0         0         0         0         0         0         0         0         0         0         0      
   0
...
...
Phase 11 will be used for (OH4,VFAT6)
...
...
----------OH4----------
VFAT0.SYNC_ERR_CNT 0x0L
VFAT1.SYNC_ERR_CNT 0x0L
VFAT2.SYNC_ERR_CNT 0x0L
VFAT3.SYNC_ERR_CNT 0x0L
VFAT4.SYNC_ERR_CNT 0x0L
VFAT5.SYNC_ERR_CNT 0x0L
VFAT6.SYNC_ERR_CNT 0xfL
VFAT7.SYNC_ERR_CNT 0x0L
VFAT8.SYNC_ERR_CNT 0x0L
```

Here the phase chosen, 11, for VFAT6 is actually bad with only 22 successes out of 50.  So this method doesn't seem to be fool proof :(

jsturdy commented:
> Here the phase chosen, 11, for VFAT6 is actually bad with only 22 successes out of 50. So this method doesn't seem to be fool proof :(

Indeed, but looking at that data, what phase would you set? (Hint, I'd toss the VFAT :-D)

I'll take this data and play with things, one thing that can immediately have an effect is the size of the search window (currently 4, but could reasonably be set to 6 or 7)
* 4: phase set to 11 (based on a bad phase of 7)
* 5: phase set to 11 (based on a bad phase of 7)
* 6: phase set to 10 (based on a bad phase of 6)
* 7: phase set to 9 (based on a bad phase of 5)
* 8: phase set to 12 (based on a bad phase of 0)

An other thing that could be done could be instead of checking `isPhaseGood(phase)`, checking if `phaseCounts[phase]==expected`, telling you instead of a one-off check of a phase, whether during the scan, that phase was "good"
* this results in setting the phase to 10 from a bad phase of 6

bdorney commented:
> I'd toss the VFAT

Yeah...we tested three different VFATs, pretty sure the problem is the GEB.

Anyway I’ll take a look at the proposal tomorrow.

jsturdy commented:
A bit more info:
* With the original algorithm, here is what the phases would be set to `[setPhase, basePhase]`:
```
Ideal phase for VFAT0 is:  [6, 8] [50. 50. 50.  0. 50. 50. 50. 50. 50. 50. 50.  0. 50. 50. 50.  0.]
Ideal phase for VFAT1 is:  [2, 4] [50. 50. 50. 50. 50.  9. 50. 50. 50. 50. 50. 50. 50.  0. 50.  0.]
Ideal phase for VFAT2 is:  [2, 4] [50. 50. 50. 50. 50. 50. 25. 50. 50. 50. 50. 50. 50. 50.  0.  0.]
Ideal phase for VFAT3 is:  [3, 5] [ 0. 50. 50. 50. 50. 50. 50. 50. 50.  0. 50. 50. 50. 50. 50.  0.]
Ideal phase for VFAT4 is:  [2, 4] [50. 50. 50. 50. 50. 50. 50. 50. 50.  0. 50. 50. 50. 50. 50.  0.]
Ideal phase for VFAT5 is:  [3, 5] [ 0. 50. 50. 50. 50. 50. 50. 50.  0. 50. 50. 50. 50. 50. 50.  0.]
Ideal phase for VFAT6 is:  [9, 10] [ 0. 50. 50. 21.  0.  0.  0.  0. 50. 50. 50. 25. 22.  0.  0.  0.]
Ideal phase for VFAT7 is:  [2, 4] [50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50.  0. 50. 50.  0.]
Ideal phase for VFAT8 is:  [6, 8] [50. 50. 50.  0. 50. 50. 50. 50. 50. 50.  0.  0. 50. 50. 50.  0.]
Ideal phase for VFAT9 is:  [2, 4] [50. 50. 50. 50. 50.  0. 50. 50. 50. 50. 50. 50. 50.  0. 50.  0.]
Ideal phase for VFAT10 is:  [3, 5] [ 0. 50. 50. 50. 50. 50. 50. 50.  0. 50. 50. 50. 50. 50. 50.  0.]
Ideal phase for VFAT11 is:  [4, 6] [50. 45. 50. 50. 50. 50. 50. 50. 50.  0. 50. 50. 50. 50. 50.  0.]
Ideal phase for VFAT12 is:  [5, 7] [50. 50.  0. 50. 50. 50. 50. 50. 50. 50.  0. 50. 50. 50. 50.  0.]
Ideal phase for VFAT13 is:  [2, 4] [50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50.  0.]
Ideal phase for VFAT14 is:  [2, 4] [50. 50. 50. 50. 50. 50.  0. 50. 50. 50. 50. 50. 50. 50.  0.  0.]
Ideal phase for VFAT15 is:  [2, 4] [50. 50. 50. 50. 50. 50. 50. 50. 50. 50.  0. 50. 50. 50. 50.  0.]
Ideal phase for VFAT16 is:  [2, 4] [50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 48. 50.  0.]
Ideal phase for VFAT17 is:  [2, 4] [50. 50. 50. 50. 50.  0. 50. 50. 50. 50. 50. 50. 50.  0. 50.  0.]
Ideal phase for VFAT18 is:  [7, 9] [50. 50. 50. 50.  0. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50.  0.]
Ideal phase for VFAT19 is:  [5, 7] [50. 50.  0. 50. 50. 50. 50. 50. 50. 50.  0. 50. 50. 50. 50.  0.]
Ideal phase for VFAT20 is:  [2, 4] [50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50.  0.]
Ideal phase for VFAT21 is:  [2, 4] [50. 50. 50. 50. 50.  0. 50. 50. 50. 50. 50. 50. 50.  0. 50.  0.]
Ideal phase for VFAT22 is:  [3, 5] [ 0. 50. 50. 50. 50. 50. 50. 50.  0. 50. 50. 50. 50. 50. 50.  0.]
Ideal phase for VFAT23 is:  [6, 8] [50. 50. 50.  0. 50. 50. 50. 50. 50. 50. 50.  0. 50. 50. 50.  0.]
```

* With the first algorithm, here is what the phases would be set to `[setPhase, basePhase(s)]`:
```
Ideal phase for VFAT0 is:  [7, [3, 11]] [50. 50. 50.  0. 50. 50. 50. 50. 50. 50. 50.  0. 50. 50. 50.  0.]
Ideal phase for VFAT1 is:  [9, [5, 13]] [50. 50. 50. 50. 50.  9. 50. 50. 50. 50. 50. 50. 50.  0. 50.  0.]
Ideal phase for VFAT2 is:  [10, [6, 14]] [50. 50. 50. 50. 50. 50. 25. 50. 50. 50. 50. 50. 50. 50.  0.  0.]
Ideal phase for VFAT3 is:  [4, [0, 9]] [ 0. 50. 50. 50. 50. 50. 50. 50. 50.  0. 50. 50. 50. 50. 50.  0.]
Ideal phase for VFAT4 is:  [5, 9] [50. 50. 50. 50. 50. 50. 50. 50. 50.  0. 50. 50. 50. 50. 50.  0.]
Ideal phase for VFAT5 is:  [4, [0, 8]] [ 0. 50. 50. 50. 50. 50. 50. 50.  0. 50. 50. 50. 50. 50. 50.  0.]
Ideal phase for VFAT6 is:  [-1, -1] [ 0. 50. 50. 21.  0.  0.  0.  0. 50. 50. 50. 25. 22.  0.  0.  0.]
Ideal phase for VFAT7 is:  [8, 12] [50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50.  0. 50. 50.  0.]
Ideal phase for VFAT8 is:  [8, [1, 15]] [50. 50. 50.  0. 50. 50. 50. 50. 50. 50.  0.  0. 50. 50. 50.  0.]
Ideal phase for VFAT9 is:  [9, [5, 13]] [50. 50. 50. 50. 50.  0. 50. 50. 50. 50. 50. 50. 50.  0. 50.  0.]
Ideal phase for VFAT10 is:  [4, [0, 8]] [ 0. 50. 50. 50. 50. 50. 50. 50.  0. 50. 50. 50. 50. 50. 50.  0.]
Ideal phase for VFAT11 is:  [5, [1, 9]] [50. 45. 50. 50. 50. 50. 50. 50. 50.  0. 50. 50. 50. 50. 50.  0.]
Ideal phase for VFAT12 is:  [6, [2, 10]] [50. 50.  0. 50. 50. 50. 50. 50. 50. 50.  0. 50. 50. 50. 50.  0.]
Ideal phase for VFAT13 is:  [-1, -1] [50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50.  0.]
Ideal phase for VFAT14 is:  [10, [6, 14]] [50. 50. 50. 50. 50. 50.  0. 50. 50. 50. 50. 50. 50. 50.  0.  0.]
Ideal phase for VFAT15 is:  [6, 10] [50. 50. 50. 50. 50. 50. 50. 50. 50. 50.  0. 50. 50. 50. 50.  0.]
Ideal phase for VFAT16 is:  [9, 13] [50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 48. 50.  0.]
Ideal phase for VFAT17 is:  [9, [5, 13]] [50. 50. 50. 50. 50.  0. 50. 50. 50. 50. 50. 50. 50.  0. 50.  0.]
Ideal phase for VFAT18 is:  [0, 4] [50. 50. 50. 50.  0. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50.  0.]
Ideal phase for VFAT19 is:  [6, [2, 10]] [50. 50.  0. 50. 50. 50. 50. 50. 50. 50.  0. 50. 50. 50. 50.  0.]
Ideal phase for VFAT20 is:  [-1, -1] [50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50.  0.]
Ideal phase for VFAT21 is:  [9, [5, 13]] [50. 50. 50. 50. 50.  0. 50. 50. 50. 50. 50. 50. 50.  0. 50.  0.]
Ideal phase for VFAT22 is:  [4, [0, 8]] [ 0. 50. 50. 50. 50. 50. 50. 50.  0. 50. 50. 50. 50. 50. 50.  0.]
Ideal phase for VFAT23 is:  [7, [3, 11]] [50. 50. 50.  0. 50. 50. 50. 50. 50. 50. 50.  0. 50. 50. 50.  0.]
```
* With the algorithm suggested in this PR (after adding the check on the set phase from the scan):
```
Ideal phase for VFAT0 is:  [7, 3] [50. 50. 50.  0. 50. 50. 50. 50. 50. 50. 50.  0. 50. 50. 50.  0.]
Ideal phase for VFAT1 is:  [9, 5] [50. 50. 50. 50. 50.  9. 50. 50. 50. 50. 50. 50. 50.  0. 50.  0.]
Ideal phase for VFAT2 is:  [10, 6] [50. 50. 50. 50. 50. 50. 25. 50. 50. 50. 50. 50. 50. 50.  0.  0.]
Ideal phase for VFAT3 is:  [4, 0] [ 0. 50. 50. 50. 50. 50. 50. 50. 50.  0. 50. 50. 50. 50. 50.  0.]
Ideal phase for VFAT4 is:  [13, 9] [50. 50. 50. 50. 50. 50. 50. 50. 50.  0. 50. 50. 50. 50. 50.  0.]
Ideal phase for VFAT5 is:  [4, 0] [ 0. 50. 50. 50. 50. 50. 50. 50.  0. 50. 50. 50. 50. 50. 50.  0.]
Ideal phase for VFAT6 is:  [10, 6] [ 0. 50. 50. 21.  0.  0.  0.  0. 50. 50. 50. 25. 22.  0.  0.  0.]
Ideal phase for VFAT7 is:  [8, 12] [50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50.  0. 50. 50.  0.]
Ideal phase for VFAT8 is:  [7, 3] [50. 50. 50.  0. 50. 50. 50. 50. 50. 50.  0.  0. 50. 50. 50.  0.]
Ideal phase for VFAT9 is:  [9, 5] [50. 50. 50. 50. 50.  0. 50. 50. 50. 50. 50. 50. 50.  0. 50.  0.]
Ideal phase for VFAT10 is:  [4, 0] [ 0. 50. 50. 50. 50. 50. 50. 50.  0. 50. 50. 50. 50. 50. 50.  0.]
Ideal phase for VFAT11 is:  [5, 1] [50. 45. 50. 50. 50. 50. 50. 50. 50.  0. 50. 50. 50. 50. 50.  0.]
Ideal phase for VFAT12 is:  [6, 2] [50. 50.  0. 50. 50. 50. 50. 50. 50. 50.  0. 50. 50. 50. 50.  0.]
Ideal phase for VFAT13 is:  [-1, -1] [50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50.  0.]
Ideal phase for VFAT14 is:  [10, 6] [50. 50. 50. 50. 50. 50.  0. 50. 50. 50. 50. 50. 50. 50.  0.  0.]
Ideal phase for VFAT15 is:  [14, 10] [50. 50. 50. 50. 50. 50. 50. 50. 50. 50.  0. 50. 50. 50. 50.  0.]
Ideal phase for VFAT16 is:  [9, 13] [50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 48. 50.  0.]
Ideal phase for VFAT17 is:  [9, 5] [50. 50. 50. 50. 50.  0. 50. 50. 50. 50. 50. 50. 50.  0. 50.  0.]
Ideal phase for VFAT18 is:  [8, 4] [50. 50. 50. 50.  0. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50.  0.]
Ideal phase for VFAT19 is:  [6, 2] [50. 50.  0. 50. 50. 50. 50. 50. 50. 50.  0. 50. 50. 50. 50.  0.]
Ideal phase for VFAT20 is:  [-1, -1] [50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50. 50.  0.]
Ideal phase for VFAT21 is:  [9, 5] [50. 50. 50. 50. 50.  0. 50. 50. 50. 50. 50. 50. 50.  0. 50.  0.]
Ideal phase for VFAT22 is:  [4, 0] [ 0. 50. 50. 50. 50. 50. 50. 50.  0. 50. 50. 50. 50. 50. 50.  0.]
Ideal phase for VFAT23 is:  [7, 3] [50. 50. 50.  0. 50. 50. 50. 50. 50. 50. 50.  0. 50. 50. 50.  0.]
```

This local testing doesn't apply the LUT, so any case 3 VFATs will report "no good phase found" for both of the new algorithms.

bdorney commented:
What would you like us to conclude from the three examples given above?

jsturdy commented:
> What would you like us to conclude from the three examples given above?

I would conclude that if the LUT for case 3 is implemented, then combined with the algorithm proposed in this PR, including latest modification to catch cases where the phase would be set to a value that was reported as bad during the scan, is so the most robust (assuming that you don't trust the original algorithm)

A failure of this algorithm should indicate that there really is a problem with setting the phase, but on the other hand, it **will** successfully find **a** phase for `VFAT6` above, which might be undesirable.
In fact, I checked a contrived case where from `VFAT6` I moved more and more of the phases to `bad` and eventually the algorithm will fail, but this is **only** because in the search, it uses a `tmpsum > sum` check, rather than a `tmpsum >= sum`.  Changing this *would* allow this algorithm to set a "good" phase in the extremely contrived case that there exists *exactly* one good phase.
In this (and similar) case(s), a check on the number of bad phase points could be implemented, and a cut if the VFAT is above this.
The only question I would leave would be whether there are extreme cases that we can contrive, but have not yet encountered, for which we would like (not) to set a good phase?

bdorney commented:
> > What would you like us to conclude from the three examples given above?
> 
> I would conclude that if the LUT for case 3 is implemented, then combined with the algorithm proposed in this PR, including latest modification to catch cases where the phase would be set to a value that was reported as bad during the scan, is so the most robust (assuming that you don't trust the original algorithm)

Okay; has this been tested already? e.g. on the coffin? I could deploy it in the lab if so.

> 
> A failure of this algorithm should indicate that there really is a problem with setting the phase, but on the other hand, it **will** successfully find **a** phase for `VFAT6` above, which might be undesirable.

Yes in this case there really was a hardware problem.

> In fact, I checked a contrived case where from `VFAT6` I moved more and more of the phases to `bad` and eventually the algorithm will fail, but this is **only** because in the search, it uses a `tmpsum > sum` check, rather than a `tmpsum >= sum`. Changing this _would_ allow this algorithm to set a "good" phase in the extremely contrived case that there exists _exactly_ one good phase.
> In this (and similar) case(s), a check on the number of bad phase points could be implemented, and a cut if the VFAT is above this.

I think if there are more than 4 bad phases we should raise some kind of flag and consider it a hardware problem/failure.  This comes from the idea that the two bad phases at most could be inbetween two phases points and they are separated.

I think this would then stop the case where VFAT6 is assigned a phase and we plow forward.

> The only question I would leave would be whether there are extreme cases that we can contrive, but have not yet encountered, for which we would like (not) to set a good phase?

I think a hard limit on the number of bad phases should be set; we could readdress this as we get more data.

jsturdy commented:
> Okay; has this been tested already? e.g. on the coffin? I could deploy it in the lab if so.

Only ex post facto on data from other scans but for the purpose of this algorithm, that should be sufficient.

> I think a hard limit on the number of bad phases should be set; we could readdress this as we get more data.

Latest push implements a variable `MAX_BAD_PHASES` that is currently set to 4.

I would feel comfortable deploying this, so if you want to do that, and give feedback, I can then clean out the previous algorithm block.

bdorney commented:
> > I think a hard limit on the number of bad phases should be set; we could readdress this as we get more data.
> 
> Latest push implements a variable `MAX_BAD_PHASES` that is currently set to 4.

Sounds good.

> 
> I would feel comfortable deploying this, so if you want to do that, and give feedback, I can then clean out the previous algorithm block.

Okay, let me know when you're remote is updated and I can `fetch` and build an updated package.

jsturdy commented:
> Okay, let me know when you're remote is updated and I can `fetch` and build an updated package.

should be with `8203832`

bdorney commented:
tested this on QC8, it's good 2 go.
